RecoLocalCalo/EcalRecProducers/plugins/EcalDetailedTimeRecHitProducer.cc

```diff
@@ -0,0 +1,246 @@
+{
+        using namespace edm;
+        using namespace reco;
+
+	edm::ESHandle<CaloGeometry>               hGeometry   ;
+	es.get<CaloGeometryRecord>().get( hGeometry ) ;
+	
+	m_geometry = &*hGeometry;
+
+        Handle< EBRecHitCollection > pEBRecHits;
+        Handle< EERecHitCollection > pEERecHits;
+
+        const EBRecHitCollection*  EBRecHits = 0;
+        const EERecHitCollection*  EERecHits = 0; 
+
+	evt.getByToken( EBRecHitCollection_, pEBRecHits);
+	if ( pEBRecHits.isValid() ) {
+	  EBRecHits = pEBRecHits.product(); // get a ptr to the product
+#ifdef DEBUG
+	  LogDebug("EcalRecHitDebug") << "total # EB rechits to be re-calibrated: " << EBRecHits->size();
+#endif
+	}
+        
+	evt.getByToken( EERecHitCollection_, pEERecHits);
+	if ( pEERecHits.isValid() ) {
+	  EERecHits = pEERecHits.product(); // get a ptr to the product
+#ifdef DEBUG
+	  LogDebug("EcalRecHitDebug") << "total # EE uncalibrated rechits to be re-calibrated: " << EERecHits->size();
+#endif
+	} 
+        
+        Handle< EcalTimeDigiCollection > pEBTimeDigis;
+        Handle< EcalTimeDigiCollection > pEETimeDigis;
+
+        const EcalTimeDigiCollection* ebTimeDigis =0;
+        const EcalTimeDigiCollection* eeTimeDigis =0;
+
+	evt.getByToken( ebTimeDigiCollection_, pEBTimeDigis);
+	//evt.getByToken( digiProducer_, pEBTimeDigis);
+	if ( pEBTimeDigis.isValid() ) {
+	  ebTimeDigis = pEBTimeDigis.product(); // get a ptr to the produc
+	  edm::LogInfo("EcalDetailedTimeRecHitInfo") << "total # ebTimeDigis: " << ebTimeDigis->size() ;
+	} 
+
+	evt.getByToken( eeTimeDigiCollection_, pEETimeDigis);
+	//evt.getByToken( digiProducer_, pEETimeDigis);
+	if ( pEETimeDigis.isValid() ) {
+	  eeTimeDigis = pEETimeDigis.product(); // get a ptr to the product
+	  edm::LogInfo("EcalDetailedTimeRecHitInfo") << "total # eeTimeDigis: " << eeTimeDigis->size() ;
+	}         
+        // collection of rechits to put in the event
+        std::auto_ptr< EBRecHitCollection > EBDetailedTimeRecHits( new EBRecHitCollection );
+        std::auto_ptr< EERecHitCollection > EEDetailedTimeRecHits( new EERecHitCollection );
+
+	GlobalPoint* vertex=0;
+
+	if (correctForVertexZPosition_)
+	  {
+	    if (!useMCTruthVertex_)
+	      {
+	      //Get the first reco vertex
+	      // get primary vertices
+		
+		edm::Handle<VertexCollection> VertexHandle;
+		evt.getByToken(recoVertex_, VertexHandle);
+		
+		if ( VertexHandle.isValid() )
+		  {
+		    if ((*VertexHandle).size()>0) //at least 1 vertex
+		      {
+			const reco::Vertex* myVertex= &(*VertexHandle)[0];
+			vertex=new GlobalPoint(myVertex->x(),myVertex->y(),myVertex->z());
+		      }
+		  }
+	
+	      }
+	    else
+	      {
+		edm::Handle<SimVertexContainer> VertexHandle;
+		evt.getByToken(simVertex_, VertexHandle);
+		
+		if ( VertexHandle.isValid() )
+		  {
+		    if ((*VertexHandle).size()>0) //at least 1 vertex
+		      {
+			assert ((*VertexHandle)[0].vertexId() == 0);
+			const SimVertex* myVertex= &(*VertexHandle)[0];
```

Comment:
It is somewhat unavoidable given the state that timing studies are in right now (experimental and constantly evolving). As things settle down this can be fixed up.